### ISSUE TITLE
New algorithm for creating menus

### ISSUE THREAD

**Issue by heatherleaf** [MUSTE-Project/MULLE#19]
I've been pondering menus, selections, linearisations, etc., and I think I have a nice and clean solution. 

First, some basic functions and types:

```
type Tokn = String
type Node = String
type Selection = [(Int, Int)]

parseSentence :: Context -> String -> [TTree]
parseSentence ctxt sent = Grammar.parseSentence (ctxtGrammar ctxt) (ctxtLang ctxt) sent

linTree :: Context -> TTree -> ([Tokn], [Node])
linTree ctxt tree = (map ltlin lintokens, map (lookupNode tree . ltpath) lintokens)
    where Linearization lintokens = Linearization.linearizeTree ctxt tree

lookupNode :: TTree -> Path -> Node
lookupNode tree path = case Tree.selectNode tree path of
                         Just (TNode node _ _) -> node

similarTrees :: Context -> TTree -> [TTree]
similarTrees ctxt tree = [ tree' | (_path, simtrees) <- Map.toList simmap, (_, tree') <- Set.toList simtrees ]
    where simmap = Prune.replaceTrees (ctxtGrammar ctxt) (ctxtPrecomputed ctxt) tree
```

Notes:
- I use `[(Int,Int)]` as selections, since that's what my implementation returns 
- I don't use LinTokens, but instead I split the linearisation into `[Tokn]` and `[Node]`
- `Node` should optimally be a pointer to a specific node in a tree
    - it must be stable: when the tree is changed (nodes inserted, deleted, moved), all remaining nodes should be treated equal
    - this includes if the node has moved, so paths are not a good implementation of nodes
- I'm using Herbert's approximation: 
    - `type Node = String` = function names
    - this makes equality over nodes "overgenerating", but it is stable
- `similarTrees` just merges all similar trees together in one large list -- it doesn't care about which path it originates from

```
getMenuItems :: Context -> String -> Map Selection (Set (Selection, [Tokn]))
getMenuItems ctxt sentence = Map.fromListWith (Set.union) $ do
    oldtree <- parseSentence ctxt sentence
    let (oldwords, oldnodes) = linTree ctxt oldtree
    newtree <- similarTrees ctxt oldtree
    let (newwords, newnodes) = linTree ctxt newtree
    let edits = alignSequences oldnodes newnodes
    let (oldselection, newselection) = splitAlignment edits
    return (oldselection, Set.singleton (newselection, newwords))
```

I haven't tested `getMenuItems`, but I hope the idea is clear enough.

The following functions are tested. `alignSequences` is a modification of an implementation of the Longest Common Subsequence algorithm, stolen from here: https://rosettacode.org/wiki/Longest_common_subsequence#Haskell

```
type Edit a = (([a], Int, Int), ([a], Int, Int))

splitAlignments :: [Edit a] -> (Selection, Selection)
splitAlignments [] = ([], [])
splitAlignments (((_,i,j), (_,k,l)) : edits) = ((i,j):oldselection, (k,l):newselection)
    where (oldselection, newselection) = splitAlignments edits

alignSequences :: Ord a => [a] -> [a] -> [Edit a]
alignSequences xs ys = mergeEdits $ snd $ a Array.! (0,0)
    where n = length xs
          m = length ys
          a = Array.array ((0,0),(n,m)) $ l1 ++ l2 ++ l3
          l1 = [((i,m), (0, [])) | i <- [0..n]]
          l2 = [((n,j), (0, [])) | j <- [0..m]]
          l3 = [((i,j), f x y i j) | (x,i) <- zip xs [0..], (y,j) <- zip ys [0..]]
          f x y i j 
              | x == y     = (eqs+1, (([x],i,i+1), ([x],j,j+1)) : eqedits)
              | ins == del = (ins,   (([x],i,i+1), ([y],j,j+1)) : eqedits)
              | ins > del  = (ins,   (([],i,i), ([y],j,j+1)) : inedits)
              | otherwise  = (del,   (([x],i,i+1), ([],j,j)) : dledits)
              where (eqs, eqedits) = a Array.! (i+1,j+1)
                    (ins, inedits) = a Array.! (i,j+1)
                    (del, dledits) = a Array.! (i+1,j)

mergeEdits :: Ord a => [Edit a] -> [Edit a]
mergeEdits [] = []
mergeEdits (edit@((xs,_,_), (xs',_,_)) : edits)
    | xs == xs' = mergeEdits edits
    | otherwise = merge edit edits
    where merge edit [] = [edit]
          merge edit@((xs,i,j), (xs',i',j')) (((ys,k,l), (ys',k',l')) : edits)
              | ys == ys' = edit : mergeEdits edits
              | otherwise = merge ((xs++ys,i,l), (xs'++ys',i',l')) edits
```

`alignSequences` creates minimal "edits" -- 1-word insertion, 1-word deletion, 1-word replacement, or 1-word don't-change-at-all. `mergeEdits` then merges edits into larger "chunks", and skips all don't-change edits (because they are not real edits).

E.g., this is what `alignSequences` (without `mergeEdits`) returns for the strings "abcdef" and "acbcxfyz":

```
  ("a",0,1) -> ("a",0,1)
  ("",1,1) -> ("c",1,2)
  ("b",1,2) -> ("b",2,3)
  ("c",2,3) -> ("c",3,4)
  ("d",3,4) -> ("x",4,5)
  ("e",4,5) -> ("",5,5)
  ("f",5,6) -> ("f",5,6)
```

And this is what `mergeEdits` then returns:

```
  ("",1,1) -> ("c",1,2)
  ("de",3,5) -> ("x",4,5)
```

Finally, from that information we can create selections for the old and new sentences.

**Comment by heatherleaf** [comment]
This would make #15 obsolete.

**Comment by heatherleaf** [comment]
How should the API to the ui look like now?

From the client to the server, it's easy: The only thing that the client should send is a string (the 2nd in the menu item). 

From the server to the client, something like this:

```
{ a: { sentence: ["flickan", "slår", "på", "en", "dator"],
       nodes: [["girl_N"], ["hit_V","switch_on_V"], ["on_Prep","switch_on_V"], ["aSg_Det"], ["computer_N"]]
       menu: ...(like above)... },
  b: { sentence: ["the", "girl", "switches", "on", "the", "computer"],
       nodes: [["girl_N"], ["switch_on_V"], ["switch_on_V"], ["theSg_Det"], ["computer_N"]]
       menu: ...(like above)... } }
```

The `nodes` can be used for highlighting. E.g., "slår" and "switches" should be highlighted, because the node "switch_on_V" is common in both A and B.

**Comment by heatherleaf** [comment]
@fredefox Any questions? Are my ideas crystal clear?

**Comment by fredefox** [comment]
Here's my first attempt. Please see caveats mentioned in the [back log](https://github.com/MUSTE-Project/MULLE/blob/d9fdd95c61e6ecaa29fc888e813af189f4f86b52/BACKLOG.md#issue-19-new-algorithm-for-creating-menus).

**Comment by heatherleaf** [comment]
Thanks! I will look through it today.

First comment: you write in the backlog that unrelated words are highlighted - I assume you mean that both occurrences of "a" are matched with "en", and both "king" with "kung". Yes it's a problem, but the only way (afaik) to get it to disappear is to use node references instead of function names. Let's go with this one, and it's not a big issue.

Second comment: your new names for linearisations are misleading (Ambiguous, Unambiguous). The `Unambiguous` linearisations (and tokens) might still be ambiguous (if `classes` has more than one element), and an `Ambiguous` linearisation can very well be unambiguous (if it has one parse tree).

I suggest to change the names:
- `Ambiguous` --> `Token.Plain` and `Linearization.Plain`
- `Unambiguous` --> `Token.Annotated` and `Linearization.Annotated`

(alternatives to Annotated can be `Labeled` or `Tagged`)

**Comment by fredefox** [comment]
Yeah I know. The names were poorly choosen. It could be (un-)annotated in stead.

**Comment by fredefox** [comment]
Note that this is not the only poorly chosen name. Also:

    module Muste.Menu.New
      ( NewFancyMenu
      , getNewFancyMenu
      )

**Comment by fredefox** [comment]
The idea was to try and re-use functionality for the almost identical types `Annotated` and `Unannotated`. Type classes are a very common pattern for this, yes.

**Comment by fredefox** [comment]
But you don't need to look through all those modules. You just need to look at what is exported. Namely:

    - Muste.Sentence
    - Muste.Sentence.Annotated
    - Muste.Sentence.Unannotated

The first one defines the interface, and the second two provides the implementations as well as some additional operations specific to those.

**Comment by fredefox** [comment]
I can recommend doing `stack haddock --open muste` to get a quick overview of how the API looks like.

**Comment by heatherleaf** [comment]
It would be much easier for me to test the new implementation if you could get the CLI to work. I tried but ran into type errors I didn't understand. 

And when you're into it: would it be possible to make the menus (and the original linearization) show the associated nodes too? Like this:

```
Sentence is now:  flickan slår på en dator    [girl_N  hit_V/switch_on_V  on_Prep/switch_on_V  aSg_Det  computer_N]

[0-0] <> flickan slår på en dator
<en> flicka slår på en dator    [aSg_Det  girl_N  hit_V/switch_on_V  on_Prep/switch_on_V  aSg_Det  computer_N]
...

[0-1] <flickan> slår på en dator
<pojken> slår på en dator    [boy_N  hit_V/switch_on_V  on_Prep/switch_on_V  aSg_Det  computer_N]
```

(I know, there are some more fancy things there :)

**Comment by fredefox** [comment]
But of course if you don't think it's an improvement, it's perfectly possible to go with two distinct data-types.

**Comment by heatherleaf** [comment]
I'm sure it's an improvement, I just didn't get it at a quick look. But I'll look into the API

**Comment by fredefox** [comment]
When you say "get the CLI' to work, you really mean switch to using `getNewFancyMenu`, right?

**Comment by fredefox** [comment]
BTW. If you're getting a lot of *warnings* (nb: not errors) then it's because I've added

    OPTIONS_GHC -Wall

To the top of most modules to enforce good coding practice and to enable some warnings that really should be the default. Like warning you on partial pattern matches which is a real no-no and name shadowing, unused variables and this sort of thing.

You can temporarily disable it while developing (preppend `--` to that line) but you really shouldn't commit anything with those errors.

**Comment by heatherleaf** [comment]
yes, I mean switch to `getNewFancyMenu`

**Comment by heatherleaf** [comment]
Btw, I just realised I forgot to push my changes to the grammar. But now finally the sentence "flickan slår på en dator" is ok!

**Comment by fredefox** [comment]
> Btw, I just realised I forgot to push my changes to the grammar. But now finally the sentence "flickan slår på en dator" is ok!

That's great! I was trying to build the grammar yesterday but was getting an out-of-memory exception from `gf`. It still spikes up to using around half of the memory on my 8 GB machine when I build now, but at least it succeeds. Pretty sure there's a space leak in gf somewhere....

**Comment by fredefox** [comment]
I've updated the cli to use the new menu algorithm. I have implemented a `Pretty`-instance for those menus, which is not so pretty... I just used the most naive implementation possible. Perhaps you can tell me how you'd like it to look.

**Comment by fredefox** [comment]
Actually 32f7177 should make the situation a bit better.

**Comment by fredefox** [comment]
Oh another thing. I'm not sure I'm using `Muste.Menu.New.Selection` correctly in all places. For instance, the change that I just pushed to the CLI asks the user to input the exact key of the menu - a `Selection` - but what I'm guessing you really want is some mechanism like the one implemented for the GUI that does some more "clever" filtering of the menus and finds all entries that somehow match. The devil is in the "somehow". For now I've just used the simplest thing I could think of which was. See the comment here:

https://github.com/MUSTE-Project/MULLE/blob/a0dfa579ae50be65db8f45fc5cd3fb707d8b77b8/muste-ajax/static/muste-gui.js#L538-L550

To sum up:

  * I've mocked the implementation in the GUI
  * The cli does no clever lookup.

**Comment by heatherleaf** [comment]
It's not necessary with the possibility to input a selection in the CLI -- it's perfectly fine that it just prints all menus for a given sentence. Like now.

But please see my new issue #20 about pretty-printing / debugging

**Comment by fredefox** [comment]
I've changed how the web ui interprets selections. I think this is correct. Prev. impl. was just quick and dirty.

https://github.com/MUSTE-Project/MULLE/commit/c6ef735489b4ec29c1acce7f264571d600730aca#diff-24dfcffbd65f7b6d8685b90da8aff880

**Comment by heatherleaf** [comment]
The new algorithm seems to work mostly fine now. There are some remaining problems, but I'll create separate issues for them instead. 

E.g. the web ui doens't work correct yet
